kernel/aster-nix/src/process/sync/condvar.rs

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::Arc;
+use core::{
+    sync::atomic::{AtomicU64, Ordering},
+    time::Duration,
+};
+
+use aster_frame::sync::{guard_lock, MutexGuard, SpinLock};
+
+use super::Pauser;
+use crate::Errno;
+
+/// The lock error type for situations where the condvar is not notified.
+pub enum LockErr<Guard> {
+    Timeout(Guard),
+    Interrupted(Guard),
+    Unknown(Guard),
+}
+/// LockResult, different from Rust std.
+/// The result of a lock operation.
+pub type LockResult<Guard> = Result<Guard, LockErr<Guard>>;
+
+impl<Guard> LockErr<Guard> {
+    pub fn into_guard(self) -> Guard {
+        match self {
+            LockErr::Timeout(guard) => guard,
+            LockErr::Interrupted(guard) => guard,
+            LockErr::Unknown(guard) => guard,
+        }
+    }
+}
+
+/// A condition variable.
+pub struct Condvar {
+    pauser: Arc<Pauser>,
+    waiter_counter: SpinLock<AtomicU64>,
+    notify_counter: SpinLock<AtomicU64>,
+}
+
+impl Condvar {
+    /// Creates a new condition variable.
+    pub fn new() -> Self {
+        Condvar {
+            pauser: Pauser::new(),
+            waiter_counter: SpinLock::new(AtomicU64::new(0)),
+            notify_counter: SpinLock::new(AtomicU64::new(0)),
+        }
+    }
+
+    /// Waits for the condition variable to be signaled or broadcasted.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> LockResult<MutexGuard<'a, T>> {
+        let lock = guard_lock(&guard);
+        drop(guard);
+        let cond = || {
+            // Check if the notify counter is greater than 0.
+            let notify_counter = self.notify_counter.lock();
```

Comment:
If you think `CondVar` will be used in interrupt context, then all locks should use `lock_irq_disabled`, otherwise dead lock may happen.